File: leetcode/array/189. Rotate Array/py/main.py
class Solution:
    # Time O(N)
    # Space O(1)
    def rotate(self, nums, k):
        n = len(nums)
        pivot = k % n

        def reverse(l, r):
            while l < r:
                nums[l], nums[r] = nums[r], nums[l]
                l, r = l + 1, r - 1

        reverse(0, n - 1)  # Reverse all entire array
        reverse(0, pivot - 1)  # Reverse left half
        reverse(pivot, n - 1)  # Reverse right half

        return nums

File: leetcode/array/189. Rotate Array/js/main.js
// Time O(N)
// Space O(1)
const rotate = (nums = [], k) => {
  k = k % nums.length;

  reverse(nums, 0, nums.length - 1);
  reverse(nums, 0, k - 1);
  reverse(nums, k, nums.length - 1);

  function reverse(nums, start, end) {
    while (start < end) {
      [nums[start], nums[end]] = [nums[end], nums[start]];
      start++;
      end--;
    }
  }
};

// Time O(N)
// Space O(N)
const rotate_II = (nums = [], k) => {
  let d = nums.length - (k % nums.length);
  let removed = nums.splice(d);
  nums.unshift(...removed);
};
